### DISCUSSION TITLE
How to correctly obtain a database connection from within DataLoaders?

### DISCUSSION THREAD

**Discussion by tmarsteel** [graphql-java/graphql-java#2374] [Q&A]
Hey guys :)

i'm running into an issue with preserving the request context when using data loaders.  
I'm serving the GraphQL queries via HTTP and every HTTP request gets a separate thread. I obtain a database connection from the pool before calling GraphQL#execute and store that connection object in ExecutionInput#context. My `DataFetcher`s can easily access it - all perfect.
But there is no such mechanism for `DataLoader`s. **How are data loaders supposed to obtain a database connection?** Especially for those that cache cross-request. This would also apply to other request context, e.g. authorization/permissions.

I have tried storing the connection object in a static `ThreadLocal`. This works for one level of data loaders, but fails with nested levels as the execution moves away from the request thread (into ForkJoinPool, via CompletableFuture.supplyAsync). For example:

```graphqls
type Country {
  id: ID!
  cities: [City!]!
}
type City {
  name: String
}
type Query {
  country(id: ID!): Country
}
schema {
  query: Query
}
```

```kotlin
object ContextTransaction {
  val entityManager = ThreadLocal<EntityManager>()
}

// in static setup code (DI container)
dataLoaderRegistry.register(
  "country",
  DataLoader.newDataLoader(
    BatchLoaderWithContext<String, Country> { key, env ->
      val entityManager = env.getContext<EntityManager>()
      // this one works, entityManager is taken from the threadlocal
      return CompletableFuture.supplyAsync {
        entityManager.findById(Country::class.java, key)
      }
    },
    DataLoaderOptions.newOptions()
      .setBatchLoaderContextProvider { ContextTransaction.entityManager.get() }
  )
)
dataLoaderRegistry.register(
  "countryCities",
  DataLoader.newMappedDataLoader(
    MappedBatchLoaderWithContext<String, List<City>> { keys, env ->
      val entityManager = env.getContext<EntityManager>()
      return CompletableFuture.supplyAsync {
        // i get an NPE here: entityManager is null
        entityManager.findCitiesByCountryIds(keys)
      }
    }
  )
)

// in the HTTP request handler, very simplified
val entityManager = entityManagerFactory.createEntityManager()
ContextTransaction.entityManager.set(entityManager);
graphQL.execute(
  ExecutionInput.newExecutionInput()
    .query(query)
    .dataLoaderRegistry(dataLoaderRegistry)
    .context(entityManager)
```

**Comment by tmarsteel** [comment]
I got it working using key contexts. Though, this was very non-obvious to me. If that *is* the correct way to do it, that should be mentioned in the docs.

**Reply by kvfedorov** [reply]
Does 'entityManager' have to be obtained from environment? Why not just pass it to constructor of dataloader?  This way you can also achieve parallel execution of dataloaders, 
as different dataloaders can use different connections...

**Reply by tmarsteel** [reply]
As i understood, caching Happens inside the dataloader. So if i create them per requests i wont get any cross-request caching. And I need it, since I have some very small, effectively invariant tables that are used in almost all requests.

**Reply by kvfedorov** [reply]
If you would initialize dataloaders (and whatever needed for them) during application startup, you should be able to reuse them between requests.
